libraries/apollo-ast/src/commonMain/kotlin/com/apollographql/apollo/ast/internal/SchemaValidationScope.kt

```diff
@@ -303,83 +358,105 @@ private fun List<GQLSchemaExtension>.getForeignSchemas(
+          prefix = urlParseResult.name
         }
 
-        val mappings = import.orEmpty().mapNotNull {
-          when (it) {
-            is GQLStringValue -> it.value to it.value
-            is GQLObjectValue -> {
-              if (it.fields.size != 2) {
-                issues.add(OtherValidationIssue("Too many fields in 'import' argument", it.sourceLocation))
-                return@mapNotNull null
-              }
-
-              val name = (it.fields.firstOrNull { it.name == "name" }?.value as? GQLStringValue)?.value
-              if (name == null) {
-                issues.add(OtherValidationIssue("import 'name' argument is either missing or not a string", it.sourceLocation))
-              }
-              val as2 = (it.fields.firstOrNull { it.name == "as" }?.value as? GQLStringValue)?.value
-              if (as2 == null) {
-                issues.add(OtherValidationIssue("import 'as' argument is either missing or not a string", it.sourceLocation))
-              }
-              if (name == null || as2 == null) {
-                return@mapNotNull null
-              }
-              name to as2
-            }
+        val import = (arguments.firstOrNull { it.name == "import" }?.value as GQLListValue?)?.values
+        val mappings = import.orEmpty().parseImport(issues)
+
+        val foreignSchema = foreignSchemas.firstOrNull { it.name == foreignName && it.version == foreignVersion }
 
-            else -> {
-              issues.add(OtherValidationIssue("Bad 'import' argument", it.sourceLocation))
-              null
+        if (foreignSchema != null) {
+          mappings.keys.forEach { key ->
+            if (foreignSchema.definitions.none { it.displayName() == key }) {
+              issues.add(
+                  OtherValidationIssue("Apollo: unknown definition '$key'", gqlDirective.sourceLocation)
+              )
             }
           }
-        }.toMap()
-
-
-        val foreignDefinitions = when (foreignName) {
-          "kotlin_labs" -> kotlinLabsDefinitions(version)
-          "nullability" -> nullabilityDefinitions(version)
-          else -> null
-        }
 
-        if (foreignDefinitions != null) {
-          val (definitions, renames) = foreignDefinitions.rename(mappings, prefix)
-          foreignSchemas.add(
-              ForeignSchema(
-                  name = foreignName,
-                  definitions = definitions,
+          val (definitions, renames) = foreignSchema.definitions.rename(mappings, prefix)
+          imports.add(
+              Import(
+                  foreignSchema = foreignSchema,
+                  renamedDefinitions = definitions,
                   newNames = renames,
-                  directivesToStrip = definitions.filterIsInstance<GQLDirectiveDefinition>().map { it.name }
               )
           )
+        } else {
+          issues.add(OtherValidationIssue("Apollo: unknown foreign schema '$foreignName/$foreignVersion'", gqlDirective.sourceLocation))
```

Comment:
Hey actually, maybe this should be allowed? In my case, the cache gradle plugin can add the cache foreign definitions 👍 - but users will still also apply the regular apollo gradle plugin, which won't accept the `extend schema @link(url: "https://specs.apollo.dev/cache/v0.1" import: ...)` with this check.

Reply:
Yikes. Good point. I still think this is something we want to validate. For an example, someone asking for `nullability/v0.1` support should get an error: this is not supported anymore. 
Yet another argument for a plugin API. Schema validation should be done only once. But plugins could contribute foreign schemas. I'll open a separate PR